eth/filters/filter.go

```diff
@@ -147,57 +147,96 @@ func (f *Filter) Logs(ctx context.Context) ([]*types.Log, error) {
 		}
 		return hdr.Number.Int64(), nil
 	}
+
+	var err error
+	// range query need to resolve the special begin/end block number
 	if f.begin, err = resolveSpecial(f.begin); err != nil {
 		return nil, err
 	}
 	if f.end, err = resolveSpecial(f.end); err != nil {
 		return nil, err
 	}
-	// Gather all indexed logs, and finish with non indexed ones
+
+	// Else fetch blcok-range logs
+	logChan, errChan := f.rangeLogsAsync(ctx, endPending)
+
+	var logs []*types.Log
+LOOP:
+	for {
+		select {
+		case log := <-logChan:
+			logs = append(logs, log)
+		case ierr := <-errChan:
+			err = ierr
+			break LOOP
+		}
+	}
+	return logs, err
+}
+
+// rangeLogsAsync retrieves block-range logs that match the filter criteria asynchronously,
+// it creates and returns two channels: one for delivering log data, and one for reporting errors.
+func (f *Filter) rangeLogsAsync(ctx context.Context, includePending bool) (chan *types.Log, chan error) {
 	var (
-		logs           []*types.Log
-		end            = uint64(f.end)
-		size, sections = f.sys.backend.BloomStatus()
+		logChan = make(chan *types.Log)
+		errChan = make(chan error)
 	)
-	if indexed := sections * size; indexed > uint64(f.begin) {
-		if indexed > end {
-			logs, err = f.indexedLogs(ctx, end)
-		} else {
-			logs, err = f.indexedLogs(ctx, indexed-1)
+
+	go func() {
```

Comment:
I haven't reviewed this in any depth, so maybe I'm missing something... but it looks to me like this spins up two routines that may be stuck forever trying to deliver, in case the other side is no longer reading, for whatever reason. 
Wouldn't it make sense it use the `ctx` to check if it's time to abort these routines?

Reply:
Good point. But seems we have handled this issue inside the `indexedLogs` https://github.com/ethereum/go-ethereum/blob/101ac683e447a2c23fc0141d33bbb3f2e0e32607/eth/filters/filter.go#L266

I wonder if I've lost something

Reply:
If we want to protect against the receiver failure, we'd have to wrap every channel send with a select, as in:

```go
select {
  case logchan <- log:
  case ctx.Done:
    return ctx.Error()
}

I guess I'm just wondering if this has an impact on performance or that should be negligible